### PR TITLE
preallocate some vec

### PR DESCRIPTION
**Pull request by klensy**
Preallocate some Vecs.

### CODE REVIEW

File: compiler/rustc_codegen_llvm/src/back/lto.rs [hunk 1]
```diff
@@ -296,6 +296,7 @@ fn fat_lto(
         // and we want to move everything to the same LLVM context. Currently the
         // way we know of to do that is to serialize them to a string and them parse
         // them later. Not great but hey, that's why it's "fat" LTO, right?
+        serialized_modules.reserve(in_memory.len());
```
Comment (pickfire) [inline compiler/rustc_codegen_llvm/src/back/lto.rs:299]: Why do we need an extra reserve here when we already did it up there?

File: compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs [hunk 2]
```diff
@@ -53,7 +53,7 @@ pub fn finalize<'ll, 'tcx>(cx: &CodegenCx<'ll, 'tcx>) {
     let mut mapgen = CoverageMapGenerator::new();
 
     // Encode coverage mappings and generate function records
-    let mut function_data = Vec::new();
+    let mut function_data = Vec::with_capacity(function_coverage_map.len());
     for (instance, function_coverage) in function_coverage_map {
```
Comment (pickfire) [inline compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs:57]: Why not do a `let function_data: Vec<_> = function_coverage_map.into_iter().map(|(instance, function_coverage)| {...}).collect();`?

File: compiler/rustc_codegen_llvm/src/back/lto.rs (hunk 1 above)
Comment (pickfire) [reply compiler/rustc_codegen_llvm/src/back/lto.rs:299]: Sorry, I thought this is `serialized_bitcode` like the above.

File: compiler/rustc_codegen_llvm/src/back/lto.rs [hunk 3]
```diff
@@ -296,6 +296,7 @@ fn fat_lto(
...
         for module in in_memory {
             let buffer = ModuleBuffer::new(module.module_llvm.llmod());
             let llmod_id = CString::new(&module.name[..]).unwrap();
```
Comment (pickfire) [inline compiler/rustc_codegen_llvm/src/back/lto.rs:302]: Should we do `sort_unstable_by_key` for the code 3 lines below this?

File: compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs (hunk 2 above)
Comment (klensy) [reply compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs:57]: This change shorter, and i don't think that converting `for` into iterator with closure in this place will give more readability or perf.

**Comment by petrochenkov** [comment]
Reserving capacities makes the logic more complex and only makes sense when there are real performance benefits.
Does this PR have measurable performance benefits?

Also many of the cases here look like candidates for using `extend`/`collect` (that should also do reservation automatically) instead of `push`es.

Reassigning because I have too many assigned PRs.
r? @bjorn3

File: compiler/rustc_codegen_llvm/src/back/write.rs [hunk 4]
```diff
@@ -329,7 +329,8 @@ unsafe extern "C" fn inline_asm_handler(diag: &SMDiagnostic, user: *const c_void
     .expect("non-UTF8 SMDiagnostic");
 
     let source = have_source.then(|| {
-        let mut spans = vec![InnerSpan::new(loc as usize, loc as usize)];
+        let mut spans = Vec::with_capacity(num_ranges + 1);
+        spans.push(InnerSpan::new(loc as usize, loc as usize));
         for i in 0..num_ranges {
```
Comment (bjorn3) [inline compiler/rustc_codegen_llvm/src/back/write.rs:334]: This is an error handling path. I don't think there is much benefit here.

**Comment by klensy** [comment]
Totally normal comment.
Try build successful

**Comment by klensy** [comment]
Closing, as this don't give any viewable boost.

**Pull request closed without merging**